test/e2e/storage/api/testsuite.go

```diff
@@ -0,0 +1,130 @@
+	tsInfo := suite.GetTestSuiteInfo()
+	testName := fmt.Sprintf("[Testpattern: %s]%s %s%s", pattern.Name, pattern.FeatureTag, tsInfo.Name, tsInfo.FeatureTag)
+	ginkgo.Context(testName, func() {
+		ginkgo.BeforeEach(func() {
+			// skip all the invalid combination of driver and pattern first
+			SkipInvalidDriverPatternCombination(driver, pattern)
+			// skip the invalid test pattern and driver combination
+			suite.skipUnsupportedTests(driver, pattern)
+		})
+		// actually define the tests
+		// at this step the testsuite should not worry about if the pattern and driver
+		// does not fit for the whole testsuite. But driver&pattern check
+		// might still needed for specific independent test cases.
+		suite.defineTests(driver, pattern)
+	})
+}
+
+// DefineTestSuites defines tests for all testpatterns and all testSuites for a driver
+func DefineTestSuites(driver TestDriver, tsInits []func() TestSuite) {
+	for _, testSuiteInit := range tsInits {
+		suite := testSuiteInit()
+		for _, pattern := range suite.GetTestSuiteInfo().TestPatterns {
+			RegisterTests(suite, driver, pattern)
+		}
+	}
+}
+
+// TestSuiteInfo represents a set of parameters for TestSuite
+type TestSuiteInfo struct {
+	Name               string              // name of the TestSuite
+	FeatureTag         string              // featureTag for the TestSuite
+	TestPatterns       []TestPattern       // Slice of TestPattern for the TestSuite
+	SupportedSizeRange e2evolume.SizeRange // Size range supported by the test suite
+}
+
+// SkipInvalidDriverPatternCombination will skip tests if the combination of driver, and testpattern
+// is not compatible to be tested. This function will be called in the TestSuiteHandler to make
+// sure all the testsuites we defined are valid.
+//
+// Whether it needs to be skipped is checked by following steps:
+// 0. Check with driver SkipUnsupportedTest
+// 1. Check if volType is supported by driver from its interface
+// 2. Check if fsType is supported
+//
+// Test suites can also skip tests inside their own skipUnsupportedTests function or in
+// individual tests.
+func SkipInvalidDriverPatternCombination(driver TestDriver, pattern TestPattern) {
+	dInfo := driver.GetDriverInfo()
+	var isSupported bool
+
+	// 0. Check with driver specific logic
+	driver.SkipUnsupportedTest(pattern)
+
+	// 1. Check if Whether volType is supported by driver from its interface
+	switch pattern.VolType {
+	case InlineVolume:
+		_, isSupported = driver.(InlineVolumeTestDriver)
+	case PreprovisionedPV:
+		_, isSupported = driver.(PreprovisionedPVTestDriver)
+	case DynamicPV, GenericEphemeralVolume:
+		_, isSupported = driver.(DynamicPVTestDriver)
+	case CSIInlineVolume:
+		_, isSupported = driver.(EphemeralTestDriver)
+	default:
+		isSupported = false
+	}
+
+	if !isSupported {
+		e2eskipper.Skipf("Driver %s doesn't support %v -- skipping", dInfo.Name, pattern.VolType)
+	}
+
+	// 2. Check if fsType is supported
+	if !dInfo.SupportedFsType.Has(pattern.FsType) {
```

Comment:
it's not your change. but for some test patterns, it does not specify a fsType which means it is default fsType. does it is considered as supported or not?

Reply:
Hmm good question. It seems when we initialize Driver we already add ""
https://github.com/kubernetes/kubernetes/blob/da75c266487c52425d3eb07df30229aea04bb28e/test/e2e/storage/drivers/csi.go#L86
The other drivers in this file also correct. So I think as long as we correctly init the driver it should all support ""

Reply:
this design means that for certain feature/capability, if no fsType is set for that feature, we can only test default fsType. For drivers that do not specify default fstype (empty string), that means this test will be skipped? 

But for snapshottable tests, I see external gcepd driver does not specify default fstype, but the tests still be triggered. https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/master/test/k8s-integration/driver-config.go#L71